wfl/fit/error.py

```diff
@@ -0,0 +1,366 @@
+                
+            label = f'{category}: {errors[category][error_type]*1e3:.2f} {_units(prop, "error")}'
+            ax_parity.scatter(ref_vals[category], pred_vals[category], label=label, edgecolors=color, facecolors='none')
+            ax_error.scatter(ref_vals[category], np.array(differences[category])*1e3, edgecolors=color, facecolors='none')
+
+        _annotate_parity_plot(ax_parity, prop, ref_property_prefix, calc_property_prefix, show_legend, error_type)
+        _annotate_error_plot(ax_error, prop, ref_property_prefix, calc_property_prefix)
+
+    plt.savefig(output, dpi=300)
+
+
+def _promote(weight, val):
+    try:
+        return weight * np.ones(val.shape)
+    except AttributeError:
+        return weight
+
+
+def _dict_add(dicts, values, at_category, prop):
+    if at_category == "_ALL_":
+        cats = [at_category]
+    else:
+        cats = [at_category, "_ALL_"]
+
+    for d, v in zip(dicts, values):
+        for cat in cats:
+            if prop not in d:
+                d[prop] = {}
+            if cat not in d[prop]:
+                d[prop][cat] = []
+            d[prop][cat].extend(v)
+
+
+def _units(prop, plt_type):
+
+    units_dict = {
+    "energy/atom": {"parity": "eV/at", "error": "meV/at"},
+    "energy": {"parity": "eV", "error": "meV"},
+    "forces": {"parity": "eV/Å", "error": "meV/Å"},
+    "virial": {"parity": "eV", "error": "meV"},
+    "virial/atom": {"parity": "eV/at", "error": "meV/at"}
+    }
+    if "virial" in prop:
+        prop = re.sub(r"/comp\b", "", prop)
+    if "forces" in prop:
+        prop = "forces"
+    try:
+        return units_dict[prop][plt_type]
+    except KeyError:
+        return "unknown units"
+
+
+def _annotate_parity_plot(ax, property, ref_property_prefix, calc_property_prefix, show_legend, error_type):
+    if show_legend:
+        ax.legend(title=f"{error_type} per category")
+    ax.set_title(f"{property} ")
+    ax.set_xlabel(f"{ref_property_prefix}{property}, {_units(property, 'parity')}")
+    ax.set_ylabel(f"{calc_property_prefix}{property}, {_units(property, 'parity')}")
+
+    xmin, xmax = ax.get_xlim()
+    ymin, ymax = ax.get_ylim()
+    lims = (min([xmin, ymin])-0.1, max([xmax, ymax])+0.1)
+    ax.set_xlim(lims)
+    ax.set_ylim(lims)
+    ax.plot(lims, lims, c='k', linewidth=0.8)
+
+    ax.grid(color='lightgrey', ls=':')
+
+
+def _annotate_error_plot(ax, property, ref_property_prefix, calc_property_prefix):
+    ax.set_title(f"{property} error")
+    ax.set_xlabel(f"{ref_property_prefix}{property}, {_units(property, 'parity')}")
+    ax.set_ylabel(f"{calc_property_prefix}{property} error, {_units(property, 'error')}") 
+    ax.grid(color='lightgrey', ls=':')
+    ax.set_yscale('log')
+
+
+def get_atomization_energy(inputs, outputs, prop_prefix, property="energy", isolated_atoms=None):
+
+    if outputs.is_done():
+        return outputs.to_ConfigSet()
+
+    isolated_atoms_from_inputs = [at for at in inputs if len(at) == 1]
```

Comment:
what's the appropriate way to find isolated atoms for periodic structures? See for one atom in a cell and check for pbc=False?

Reply:
I'm not sure, and you can tell it's hard to do robustly, because every fitting code does it differently. I'm OK with (1) checking for one atom and pbc=False, or (2) standardizing on some info field and its value, or (3) letting the user specify the field name and its value. I lean toward the 2nd of these.  The first strikes me as too likely to lead to unintended hits, and the 3rd seems overly general (although we could do 3 defaulting to 2, e.g. optional arguments, `isolated_atom_info_field="config_type"` and `isolated_atom_field_value="isolated_atom"`). We should makes the fitting wrapper functions use the same arguments, and convert the isolated atoms into the appropriate input for each actual fitting code.

However, I'm also not sure atomization energy belongs in `wfl.fit.error`. It's a generic sort of thing you might want to compute.

Reply:
I'll move the function to `wfl/utils/configs.py::atomization_energy` (any better names?). I'd like to have the flexibility for 3rd option. I think `config_type` default for `isolated_atom_info_key` would be ok (little chance to need to set isolated atom's `config_type` to anything else?). And for `isolated_atom_field_value` I'd leave it for `None` or `"default"` which would check for either `isolated_atom` (a sensible default) or `IsolatedAtom` (what MACE checks for). (Maybe worth opening a PR for MACE to be more flexible here?..)

Reply:
I'm happy with `isolated_atom_info_key="config_type", isolated_atom_field_value="default"`. We should make sure all the fitting functions that need this info use this key from here on.  

As for MACE, I could go either way. We could say that `wfl` always uses `"isolated_atom"`, and convert ourselves when calling `mace_fit`, or we could allow for both.  I tend to prefer the former.  Do you remember how ACE fit and/or GAP fir identify them, if they do at all?

Reply:
Do we want to be able to explicitly pass a dict of atomic energies to this atomization energy (which will override the configs?), for people who don't routinely run databases with isolated atom configs?

Reply:
> I'm happy with `isolated_atom_info_key="config_type", isolated_atom_field_value="default"`

great. 

> We could say that wfl always uses "isolated_atom"

Sounds good. 

> Do you remember how ACE fit and/or GAP fir identify them, if they do at all?

They don't, but I'll go over them in this fix too. 

> Do we want to be able to explicitly pass a dict of atomic energies to this atomization energy (which will override the configs?), for people who don't routinely run databases with isolated atom configs?

In principle I think that would be useful. In practice though, I think for ACE that's technically supported - just need to include them in the fitting dictionary and I think the same goes for GAP. So then I don't see why to double-up in yet another way?

Reply:
Is this all done now?  I saw some changes to `fit/ace.py`.

Reply:
yes - please resolve this if it looks ok. Also I think this is the last comment - please add more or go ahead and merge.

Reply:
Going to resolve this, since it's basically OK,  although I did start a discussion about the specific implementation.